Added a How-To for Dapr building blocks (#1045)
radius-project/docs: docs/content/guides/author-apps/dapr/how-to-dapr-building-block/index.md

---
type: docs
title: "How-To: Add a Dapr building block"
linkTitle: "Add a Dapr building block"
description: "Learn how to add a Dapr building block to a Radius Application"
weight: 300
categories: "How-To"
tags: ["Dapr"]
---

This how-to guide will provide an overview of how to:

> - Leverage a [Dapr building block](https://docs.dapr.io/developing-applications/building-blocks/) with your Radius Application

Reviewer: ```suggestion
- Leverage a [Dapr building block](https://docs.dapr.io/developing-applications/building-blocks/) in your Radius Application
```

## Prerequisites

- [Supported Kubernetes cluster]({{< ref "/guides/operations/kubernetes/overview#supported-kubernetes-clusters" >}})
- [rad CLI]({{< ref getting-started >}})
- [Dapr CLI](https://docs.dapr.io/getting-started/install-dapr-cli/)
- [Radius initialized with `rad init`]({{< ref howto-environment >}})
- [Dapr initialized with `dapr init -k`](https://docs.dapr.io/getting-started/install-dapr-selfhost/)

## Step 1: Start with a container with a Dapr sidecar

Begin by creating a file named `app.bicep` with a defined Radius container and a Dapr sidecar, if you need a in-depth walkthrough see the ['How-To: Add a Dapr sidecar to a container' guide]({{< ref how-to-dapr-sidecar >}}):

> {{< rad file="../how-to-dapr-sidecar/snippets/app-sidecar.bicep" embed=true >}}

Reviewer: Let's not take dependencies on files that aren't in this directory. This will cause issues if we ever need to move pages around later.

## Step 2: Add a Dapr state store resource

> Now add the Dapr state store resource, which will allow Dapr's state management API to save, read and query key/value pairs for resources such as a Redis Cache:

Reviewer: ```suggestion
Now add a Dapr state store resource, which models a [Dapr state store component](https://docs.dapr.io/developing-applications/building-blocks/state-management/state-management-overview/):
```

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//STATESTORE" >}}

## Step 3: Add a connection from the container resource to the Dapr state store resource

> Edit your container resource to allow the Dapr state store to establish a connection:

Reviewer: ```suggestion
Update your container resource with a connection to the Dapr state store. This will inject important connection information (the component name) into the container's environment variables:
```

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//CONTAINER" >}}

## Step 3: Deploy the application

> Deploy your application with the `Rad CLI`:

Reviewer: ```suggestion
Deploy your application with the `rad` CLI:
```

> ```

Reviewer 2: ```suggestion
```bash
```

rad run ./app.bicep -a demo
```

Your console output should look similar to:

> ```bash

Reviewer 2: remove `bash` from this snippet annotation so it renders like console output

Building app.bicep...
Deploying template 'app.bicep' for application 'demo' and environment 'default' from workspace 'default'...

Deployment In Progress... 

Completed            statestore      Applications.Dapr/stateStores
...                  demo            Applications.Core/containers

Deployment Complete

Resources:
    demo            Applications.Core/containers
    statestore      Applications.Dapr/stateStores

Starting log stream...
```

> 

Reviewer: ```suggestion
I'd add a step here to open the sample container in your browser to see the injected environment variables from Dapr and to save/get some todo items using the statestore.
```

## Step 4: Verify the Dapr statestore

> Run `kubectl get pods -n default-demo` to see all the pods running in your Kubernetes cluster:

Reviewer 2: avoid stating the command in this line if you're already including it in the bash snippet below. instead, you can say somthing like "Run the command below to..."

> ```

Reviewer 2: ```suggestion
```bash
```

> kubectl get pods -n default-demo

Reviewer: Instead of getting pods, I'd get the Dapr components. Or maybe a better solution could be to use the Dapr CLI to get the dapr components.

```

The console output should similar to:

> ```bash

Reviewer 2: remove `bash` from this snippet annotation so it renders like console output

NAME                                 READY   STATUS    RESTARTS      AGE
redis-r5tcrra3d7uh6-7679cd55-j4sxj   2/2     Running   6 (18m ago)   25h
demo-5f69fff66c-jtdwz                2/2     Running   0             10m
```

## Done

You've successfully deployed a Radius container with a Dapr sidecar and a Dapr statestore in the form of a Redis Cache!

## Cleanup

> Run the following command to delete your app and container:

Reviewer: ```suggestion

Run the following command to delete your app, including the container and the state store:
```

Reviewer 2: also recommend linking to the [delete](https://docs.radapp.io/reference/cli/rad_application_delete/) section of the docs

> ```

Reviewer 2: ```suggestion
```bash
```

---
type: docs
title: "How-To: Add a Dapr building block"
linkTitle: "Add a Dapr building block"
description: "Learn how to add a Dapr building block to a Radius Application"
weight: 300
categories: "How-To"
tags: ["Dapr"]
---

> 

Author: ```suggestion
```

---
type: docs
title: "How-To: Add a Dapr building block"

> linkTitle: "Add a Dapr building block"

Reviewer: If this wraps around dur to length you can remove the "Dapr" from the linkTitle.

description: "Learn how to add a Dapr building block to a Radius Application"
weight: 300
categories: "How-To"
tags: ["Dapr"]
---

This how-to guide will provide an overview of how to:

- Leverage a [Dapr building block](https://docs.dapr.io/developing-applications/building-blocks/) in your Radius Application

## Prerequisites

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- [Radius environment]({{< ref "installation#step-3-initialize-radius" >}})

> - [Setup a supported Kubernetes cluster](https://docs.radapp.io/guides/operations/kubernetes/overview/#supported-clusters)

Reviewer: These two are out of order. I'd remove the supported cluster piece since it's already a pre-requisite of the Radius Environment.

- [Dapr installed on your Kubernetes cluster](https://docs.dapr.io/operations/hosting/kubernetes/kubernetes-deploy/)

> ## Step 1: Start with a container with a Dapr sidecar

Reviewer: ```suggestion
## Step 1: Start with a container and a Dapr sidecar
```

Begin by creating a file named `app.bicep` with a defined Radius container and a Dapr sidecar, if you need a in-depth walkthrough see the ['How-To: Add a Dapr sidecar to a container' guide]({{< ref how-to-dapr-sidecar >}}):

{{< rad file="./snippets/app-sidecar.bicep" embed=true >}}

## Step 2: Add a Dapr state store resource

> Now add a Dapr state store resource, which models a [Dapr state store component](https://docs.dapr.io/developing-applications/building-blocks/state-management/state-management-overview/):

Reviewer: I'd add in a small blurb about this using the default Recipe in the environment. (and also double check the pre-reqs to make sure we tell the user they need local-dev Recipes)

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//STATESTORE" >}}

## Step 3: Add a connection from the container resource to the Dapr state store resource

Update your container resource with a connection to the Dapr state store. This will inject important connection information (the component name) into the container's environment variables:

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//CONTAINER" >}}

> > Note this guide leverages the Dapr state store [Radius Recipe]({{< ref "/guides/recipes/overview" >}}), the template configurations can be found in the [Radius Recipe repo](https://github.com/radius-project/recipes/tree/main).

Reviewer: Please move this further up, as this doesn't have to do with connections.

## Step 3: Deploy the application

> Deploy your application with the `rad` CLI:

Reviewer: ```suggestion
Run your application with the `rad` CLI:
```

```bash
rad run ./app.bicep -a demo
```

Your console output should look similar to:

```
Building ./app.bicep...
Deploying template './app.bicep' for application 'demo' and environment 'default' from workspace 'default'...

Deployment In Progress... 

Completed            statestore      Applications.Dapr/stateStores
...                  demo            Applications.Core/containers

Deployment Complete

Resources:
    demo            Applications.Core/containers
    statestore      Applications.Dapr/stateStores

Starting log stream...
```

> I'd add a step here to open the sample container in your browser to see the injected environment variables from Dapr and to save/get some todo items using the statestore.

Reviewer: This looks like it needs to be added still

## Step 4: Verify the Dapr statestore

Run the command below to see all the pods running in your Kubernetes cluster:

```bash
dapr components -k
```

The console output should similar to:

```

> ⚠  In future releases, this command will only query the "default" namespace by default. Please use the --namespace flag for a specific namespace, or the --all-namespaces (-A) flag for all namespaces.

Reviewer: Please update the command so it doesn't emit a warning

Author: The comment prints out regardless of usage of `--namespace` from testing it. But have update the command above.

---
type: docs
title: "How-To: Add a Dapr building block"
linkTitle: "Add a Dapr building block"
description: "Learn how to add a Dapr building block to a Radius Application"
weight: 300
categories: "How-To"
tags: ["Dapr"]
---

> -Add a Dapr state store resource

Author: ```suggestion
```

> -Add a connection from the container to the statestore

Author: ```suggestion
```

> -Deploy the application

Author: ```suggestion
```

> -Use the Dapr CLI or kubectl CLI to "prove what just happened"

Reviewer 2: Should this be listed below the overview sentence instead?

I also recommend shortening these bullet points to provide a more brief overview: use building block to deploy state store, connect container to dapr state store

> -Use the Dapr CLI or kubectl CLI to "prove what just happened"

Author: ```suggestion
```

This how-to guide will provide an overview of how to:

- Leverage a [Dapr building block](https://docs.dapr.io/developing-applications/building-blocks/) with your Radius Application

## Prerequisites

- [Supported Kubernetes cluster]({{< ref "/guides/operations/kubernetes/overview#supported-kubernetes-clusters" >}})
- [rad CLI]({{< ref getting-started >}})
- [Dapr CLI](https://docs.dapr.io/getting-started/install-dapr-cli/)
- [Radius initialized with `rad init`]({{< ref howto-environment >}})

> - [Dapr initialized with `dapr init -k`](https://docs.dapr.io/getting-started/install-dapr-selfhost/)

Reviewer 2: take a look at the prereqs here, I recommend aligning closer to this: https://docs.radapp.io/tutorials/dapr/#prerequisites

```

Your console output should look similar to:

```bash
Building app.bicep...
Deploying template 'app.bicep' for application 'demo' and environment 'default' from workspace 'default'...

Deployment In Progress... 

Completed            statestore      Applications.Dapr/stateStores
...                  demo            Applications.Core/containers

Deployment Complete

Resources:
    demo            Applications.Core/containers
    statestore      Applications.Dapr/stateStores

Starting log stream...
```

## Step 4: Verify the Dapr statestore

Run `kubectl get pods -n default-demo` to see all the pods running in your Kubernetes cluster:

```
kubectl get pods -n default-demo
```

The console output should similar to:

```bash
NAME                                 READY   STATUS    RESTARTS      AGE
redis-r5tcrra3d7uh6-7679cd55-j4sxj   2/2     Running   6 (18m ago)   25h
demo-5f69fff66c-jtdwz                2/2     Running   0             10m
```

## Done

You've successfully deployed a Radius container with a Dapr sidecar and a Dapr statestore in the form of a Redis Cache!

## Cleanup
Run the following command to delete your app and container:

```
rad app delete -a demo
```

> ## Further reaading

Author: ```suggestion
## Further reading
```

linkTitle: "Add a building block"
description: "Learn how to add a Dapr building block to a Radius Application"
weight: 300
categories: "How-To"
tags: ["Dapr"]
---

This how-to guide will provide an overview of how to:

- Leverage a [Dapr building block](https://docs.dapr.io/developing-applications/building-blocks/) in your Radius Application

## Prerequisites

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Radius Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- [Radius environment]({{< ref "installation#step-3-initialize-radius" >}})
- [Radius local-dev Recipes]({{< ref howto-dev-recipes >}})
- [Dapr installed on your Kubernetes cluster](https://docs.dapr.io/operations/hosting/kubernetes/kubernetes-deploy/)

## Step 1: Start with a container and a Dapr sidecar

Begin by creating a file named `app.bicep` with a defined Radius container and a Dapr sidecar, if you need a in-depth walkthrough see the ['How-To: Add a Dapr sidecar to a container' guide]({{< ref how-to-dapr-sidecar >}}):

{{< rad file="./snippets/app-sidecar.bicep" embed=true >}}

## Step 2: Add a Dapr state store resource

> Now add a Dapr state store resource, which models a [Dapr state store component](https://docs.dapr.io/developing-applications/building-blocks/state-management/state-management-overview/). The Dapr state store resource in this example is defined in a pre-defined Recipe from the Radius [`local-dev` Recipes pack]({{< ref "guides/recipes/overview##use-lightweight-local-dev-recipes" >}}):

Reviewer: ```suggestion
Now add a Dapr state store resource, which models a [Dapr state store component](https://docs.dapr.io/developing-applications/building-blocks/state-management/state-management-overview/). The underlying infrastructure and Dapr component configuration are deployed via a [`local-dev` Recipe]({{< ref "guides/recipes/overview##use-lightweight-local-dev-recipes" >}}), which leverages a lightweight Redis container:
```

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//STATESTORE" >}}

> > Note the template configurations can be found in the [Radius Recipe repo](https://github.com/radius-project/recipes/tree/main).

Reviewer: ```suggestion
> Visit the [Radius Recipe repo](https://github.com/radius-project/recipes/blob/main/local-dev/statestores.bicep) to learn more about `local-dev` Recipes and view the Dapr State Store Recipe used in this guide.
```

## Step 3: Add a connection from the container resource to the Dapr state store resource

Update your container resource with a connection to the Dapr state store. This will inject important connection information (the component name) into the container's environment variables:

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//CONTAINER" >}}

> 
> 

Reviewer: ```suggestion

```

## Step 3: Deploy the application

Run your application with the `rad` CLI:

```bash
rad run ./app.bicep -a demo
```

Your console output should look similar to:

```
Building ./app.bicep...
Deploying template './app.bicep' for application 'demo' and environment 'default' from workspace 'default'...

Deployment In Progress... 

Completed            statestore      Applications.Dapr/stateStores
...                  demo            Applications.Core/containers

Deployment Complete

Resources:
    demo            Applications.Core/containers
    statestore      Applications.Dapr/stateStores

Starting log stream...
```

> Your demo application should contain the following connection information:

Reviewer: Please add a step with a clickable link to instruct the user how to open the demo application.

{{< image src="app-statestore.png" alt="Screenshot of the demo Redis connection" width=700px >}}

> 
> 

Reviewer: ```suggestion

```

## Step 4: Verify the Dapr statestore

Run the command below to see all the pods running in your Kubernetes cluster:

```bash
dapr components --namespace "default-demo" -k
```

The console output should similar to:

```
NAMESPACE     NAME        TYPE         VERSION  SCOPES  CREATED              AGE  
default-demo  statestore  state.redis  v1               2024-02-19 17:13.20  2m   
```

## Done

> You've successfully deployed a Radius container with a Dapr sidecar and a Dapr statestore in the form of a Redis Cache!

Reviewer: ```suggestion
You've successfully deployed a Radius container with a Dapr sidecar along with a Dapr State Store. With the combination of Radius + Dapr, both your application's code and it's definition are now platform neutral.
```

## Cleanup

To delete your app, run the [rad app delete]({{< ref rad_application_delete >}}) command to cleanup the app and its resources, including the Recipe resources:

```bash
rad app delete -a demo
```

## Further reading

- [Dapr state management](https://docs.dapr.io/developing-applications/building-blocks/state-management/)

> - [Dapr state store resource schema]({{< ref dapr-statestore >}})

Reviewer: Instead of these links being specific to Dapr State Stores, could you make them link to Dapr Building Blocks and all of the Dapr schemas?

{{< rad file="./snippets/app-statestore.bicep" embed=true marker="//CONTAINER" >}}

## Step 3: Deploy the application

Run your application with the `rad` CLI:

```bash
rad run ./app.bicep -a demo
```

Your console output should look similar to:

```
Building ./app.bicep...
Deploying template './app.bicep' for application 'demo' and environment 'default' from workspace 'default'...

Deployment In Progress... 

Completed            statestore      Applications.Dapr/stateStores
...                  demo            Applications.Core/containers

Deployment Complete

Resources:
    demo            Applications.Core/containers
    statestore      Applications.Dapr/stateStores

Starting log stream...
```

Your demo application should contain the following connection information:

{{< image src="app-statestore.png" alt="Screenshot of the demo Redis connection" width=700px >}}

## Step 4: Verify the Dapr statestore

Run the command below to see all the pods running in your Kubernetes cluster:

```bash
dapr components --namespace "default-demo" -k
```

The console output should similar to:

```

> ⚠  In future releases, this command will only query the "default" namespace by default. Please use the --namespace flag for a specific namespace, or the --all-namespaces (-A) flag for all namespaces.

Author: ```suggestion
```
